src/adjmat/AdjacencyMatrixBase.cpp

```diff
@@ -215,52 +215,81 @@ unsigned AdjacencyMatrixBase::retrieveNeighbours( const unsigned& current, std::
+    indices.resize( 1+ablocks.size()+threeblocks.size() );
+  }
 
   // Now get the positions
-  const Value* myval = getConstPntrToComponent(0);
-  unsigned natoms=retrieveNeighbours( current, indices ); myvals.setSplitIndex( natoms );
+  // const Value* myval = getConstPntrToComponent(0);
+  unsigned natoms=retrieveNeighbours( current, indices );
+  myvals.setSplitIndex( natoms );
 
+  //placing this on the stack
+  Vector currentAtom = ActionAtomistic::getPosition(current);
   // Now retrieve everything for the third atoms
   if( threeblocks.size()>0 ) {
-    unsigned ncells_required=0; std::vector<unsigned> cells_required( threecells.getNumberOfCells() );
-    threecells.addRequiredCells( threecells.findMyCell( ActionAtomistic::getPosition(current) ), ncells_required, cells_required );
+    unsigned ncells_required=0;
+    std::vector<unsigned> cells_required( threecells.getNumberOfCells() );
+    threecells.addRequiredCells( threecells.findMyCell( currentAtom ), ncells_required, cells_required );
     threecells.retrieveAtomsInCells( ncells_required, cells_required, natoms, indices );
   }
   myvals.setNumberOfIndices( natoms );
 
 // Apply periodic boundary conditions to atom positions
-  std::vector<std::vector<Vector> > & t_atoms( myvals.getFirstAtomDerivativeVector() ); if( t_atoms.size()!=1 ) t_atoms.resize(1);
-  if( t_atoms[0].size()<getNumberOfAtoms() ) t_atoms[0].resize( getNumberOfAtoms() );
-  for(unsigned i=0; i<natoms; ++i) t_atoms[0][i] = ActionAtomistic::getPosition(indices[i]) - ActionAtomistic::getPosition(current);
-  if( !nopbc ) pbcApply( t_atoms[0], natoms );
+  std::vector<std::vector<Vector> > & t_atoms( myvals.getFirstAtomDerivativeVector() );
+  //since I am usig this as a temp variable, just have to care if size is 0,
+  if( t_atoms.size()==0 )
+    t_atoms.resize(1);
+  if( t_atoms[0].size()<getNumberOfAtoms() )
+    t_atoms[0].resize( getNumberOfAtoms() );
+  for(unsigned i=0; i<natoms; ++i)
+    t_atoms[0][i] = ActionAtomistic::getPosition(indices[i]) - currentAtom;
+  if( !nopbc )
+    pbcApply( t_atoms[0], natoms );
   // And collect atom position data
   std::vector<Vector> & atoms( myvals.getAtomVector() );
-  if( atoms.size()<getNumberOfAtoms() ) atoms.resize( getNumberOfAtoms() );
-  for(unsigned i=0; i<natoms; ++i) atoms[ indices[i] ] = t_atoms[0][i];
+  if( atoms.size()<getNumberOfAtoms() )
+    atoms.resize( getNumberOfAtoms() );
+  for(unsigned i=0; i<natoms; ++i)
+    atoms[ indices[i] ] = t_atoms[0][i];
 }
 
 void AdjacencyMatrixBase::performTask( const std::string& controller, const unsigned& index1, const unsigned& index2, MultiValue& myvals ) const {
-  Vector zero; zero.zero(); plumed_dbg_assert( index2<myvals.getAtomVector().size() );
+  Vector zero; zero.zero();
+  plumed_dbg_assert( index2<myvals.getAtomVector().size() );
   double weight = calculateWeight( zero, myvals.getAtomVector()[index2], myvals.getNumberOfIndices()-myvals.getSplitIndex(), myvals );
```

Comment:
@gtribello 
Since I do not know all the details, I have a question on `calculateWeight`: by git-grepping this is only called here.

And so the first argument is a fixed `{0.0,0.0,0.0}`, and it is only used in `HBPammMatrix` `TopologyMatrix` to do something like "`pos2-0`".
My suggestion is to change the signature to something like `double calculateWeight(Vector , unsigned , MultiValue& )` so there is no need to copy the already calculated distance in a temporary value, moreover the signature can be that, but in the definition, you can freely add `const` to variables passed by value.

I am basing this suggestion also on [this](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const))

Reply:
Yes, I think if you don't pass the first argument, it would be fine.  The pbc are all applied in one place because Michele Ceriotti wrote a fast algorithm to do pbc distance on many arguments at the same time, which is used here.